crates/wasi-common/winx/src/ntdll.rs

```diff
@@ -49,17 +51,39 @@ impl Default for IO_STATUS_BLOCK {
     }
 }
 
-#[link(name = "ntdll")]
-extern "C" {
+macro_rules! ntdll_import {
+    { fn $name:ident($($arg:ident: $argty:ty),*) -> $retty:ty; $($tail:tt)* } => {
+        pub(crate) unsafe fn $name($($arg: $argty),*) -> $retty {
+            static ADDRESS: AtomicUsize = AtomicUsize::new(0);
+            let address = match ADDRESS.load(Ordering::Relaxed) {
+                0 => {
+                    let ntdll = GetModuleHandleA("ntdll\0".as_ptr() as *const i8);
+                    let address = GetProcAddress(
+                        ntdll,
+                        concat!(stringify!($name), "\0").as_ptr() as *const i8,
+                    ) as usize;
+                    ADDRESS.store(address, Ordering::Relaxed);
```

Comment:
Could you add an assert here that `address` isn't null as well?

Reply:
Done